File: SovietTransistorsDatabase/Domain/TransistorValidator.cs
namespace SovietTransistorsDatabase.Domain;

public static class TransistorValidator
{
    public static IReadOnlyList<string> Errors(Transistor transistor)
    {
        var errors = new List<string>();
        if (!Materials.IsValidSymbol(transistor.Material))
            errors.Add($"материал «{transistor.Material}» должен быть Г/1, К/2, А/3 или И/4");
        if (transistor.Subclass is not ('Т' or 'П'))
            errors.Add($"подкласс «{transistor.Subclass}» должен быть Т или П");
        if (transistor.Feature is < 1 or > 9)
            errors.Add($"характерный эксплуатационный признак {transistor.Feature} должен быть от 1 до 9");
        if (transistor.DevelopmentNumber is < 1 or > 999)
            errors.Add($"порядковый номер разработки {transistor.DevelopmentNumber} должен быть от 1 до 999");
        if (!Cyrillic.IsUpperLetters(transistor.Letters))
            errors.Add($"классификация «{transistor.Letters}» должна быть одной или двумя заглавными русскими буквами");
        if (transistor.Modification is < 1 or > 9)
            errors.Add($"модификация {transistor.Modification} должна быть от 1 до 9");
        if (transistor.ChipVariant is < 1 or > 6)
            errors.Add($"бескорпусное исполнение {transistor.ChipVariant} должно быть от 1 до 6");
        return errors;
    }
}

File: internal/domain/transistorvalidator.go
package domain

import "fmt"

func ValidateTransistor(t Transistor) []string {
	var errors []string
	if !IsValidMaterialSymbol(t.Material) {
		errors = append(errors, fmt.Sprintf("материал «%c» должен быть Г/1, К/2, А/3 или И/4", t.Material))
	}
	if t.Subclass != 'Т' && t.Subclass != 'П' {
		errors = append(errors, fmt.Sprintf("подкласс «%c» должен быть Т или П", t.Subclass))
	}
	if t.Feature < 1 || t.Feature > 9 {
		errors = append(errors, fmt.Sprintf("характерный эксплуатационный признак %d должен быть от 1 до 9", t.Feature))
	}
	if t.DevelopmentNumber < 1 || t.DevelopmentNumber > 999 {
		errors = append(errors, fmt.Sprintf("порядковый номер разработки %d должен быть от 1 до 999", t.DevelopmentNumber))
	}
	if !IsUpperLetters(t.Letters) {
		errors = append(errors, fmt.Sprintf("классификация «%s» должна быть одной или двумя заглавными русскими буквами", t.Letters))
	}
	if t.Modification != nil && (*t.Modification < 1 || *t.Modification > 9) {
		errors = append(errors, fmt.Sprintf("модификация %d должна быть от 1 до 9", *t.Modification))
	}
	if t.ChipVariant != nil && (*t.ChipVariant < 1 || *t.ChipVariant > 6) {
		errors = append(errors, fmt.Sprintf("бескорпусное исполнение %d должно быть от 1 до 6", *t.ChipVariant))
	}
	return errors
}
